core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -292,10 +330,20 @@ private static void validate(SnapshotId snapshotId) {
      * @param partial                    allow partial snapshots
      * @param userCreateSnapshotListener listener
      */
-    private void beginSnapshot(ClusterState clusterState, final SnapshotsInProgress.Entry snapshot, final boolean partial, final CreateSnapshotListener userCreateSnapshotListener) {
+    private void beginSnapshot(final ClusterState clusterState,
+                               final SnapshotsInProgress.Entry snapshot,
+                               final boolean partial,
+                               final CreateSnapshotListener userCreateSnapshotListener) {
         boolean snapshotCreated = false;
         try {
-            Repository repository = repositoriesService.repository(snapshot.snapshotId().getRepository());
+            // check if the snapshot name already exists in the repository
+            if (snapshotNameExists(snapshot.snapshot())) {
```

Comment:
I've been trying to capture some of the race conditions I've encountered here: https://github.com/elastic/elasticsearch/issues/18551

But I think there are a few race conditions in general, and one of them relates to the fact that we have two separate places where we are retrieving snapshot info without a lock (cluster state and repository), so at any given time, the state could change and for example, a snapshot can move from in progress (cluster state) to completed (in repository), and we could double count if we don't properly synchronize around these snapshot related operations.